### ISSUE TITLE
As an administrator, I want to define a custom ordering of fields from the administrator control panel.

### ISSUE THREAD

**Issue by gorel** [gorel/dli-reports#53]

**Comment by gorel** [comment]
This is actually a fairly large task that I think we'll need Jon's help with. Here's what I'm thinking:

I'll edit the database to support unique orderings (still coming up with an ideal schema for this).

I'll work with Gunny to set up routes and edit forms appropriately.

I'm thinking that after a user hits create/edit report, they're taken to a second screen where they can click and drag the fields in the order of their choosing.

Jon will probably have to help us figure out how to make the UI side of things for this.

**Comment by faultyserver** [comment]
I can go into more detail at our meeting today, but this is my suggestion based on facing this exact task in my project from the summer.

First, a preface: I would rather see the ordering be defined on a per-report basis. Not only is this easier and more flexible for the users, but it's also actually going to be easier for us to implement. Defining a "global default" for the order that fields appear is something that I think the people at DLI would find tedious; having to leave the report they are (probably) working on, go to the administrative panel, find the field they want to move, and find where they want it to be amidst a slew of fields that are irrelevant is just bad UX. We want users to be able to do everything related to reports on the reports view, and leave the admin panel for administrative tasks like managing users, departments, etc.

In a truly ideal situation, we would even have department-specific field orderings, but that would require a lot more rewriting of tables/controllers, so I suggest we skip that.

Now, for how I would go about implementing this. In order:
1. The `Report` and `Field` models need to change from what's called a `has_and_belongs_to_many` relationship (one where the ORM takes care of the join table), to what's called a `has_many through` relationship (one where the developer can edit the structure of the table. SQLAlchemy calls this an ["Association Proxy"](http://docs.sqlalchemy.org/en/latest/orm/extensions/associationproxy.html#simplifying-association-objects), and doing this will require making a new model `ReportField` that represents the association, and that we can work with directly.
2. In the new `ReportField` model, define a new field, `index`, that is an `INT(11)`. This will represent the order that the fields are to appear on the page. (Doing department-specific ordering would require changing the table to `reports_departments_fields` and having the index represent the order that the field will appear in that department for that report. The same thing, really, but more specific).
3. In the `reports/edit` and `reports/create` forms, add an array field called `field_order`. This will get filled with `Field` ids in the order that the user has placed them. The corresponding controllers will need to set the `index` field on the `ReportField` object based on the index of the field in this array.
4. In the `reports/submit_data` controllers, when selecting the fields to send to the view, add an `.order_by(ReportField.index)` clause to the SQLAlchemy query. This will result in an array of fields that are already in order for display in the view.

At this point, we have set up the data structure for ordering the fields, and have the infrastructure for filling in and using this information. For the UI, there are two distinct sides: editing the view for entering the ordering information, and editing the view to display the fields in order.

For the latter, if the controllers are set up properly, then the views will not have to do any work (which would be great, because that is one of the primary goals of MVC), and the fields should just appear in the right order without any changes to its markup.

For the former, however, there is a fair amount of work to do. Importantly, though, we should NOT create a second view just for ordering of fields. This kind of pattern (called a Multi-Step Flow) is only applicable when there is NO chance that the user would need to go back and edit there previous entries. This does not fit our use case, as a user may have forgotten to add a field to their report, and would be forced to continue ordering fields, then go back and edit the report and re-order fields just to add the field they missed.

Instead, I'm planning on implementing all of the editing/creation tools on a single page. The changes will look something like this:
1. The large list of field options with checkboxes will be transformed into a searchable tree view of sorts on the left side of the page, keyed by department, ordered alphabetically (I can take care of all of this formatting on the front end with JavaScript, no backend changes necessary).
2. The user can select fields either by double clicking or dragging them to the right side of the page, which will be a large-format display of all of the fields that are currently in the report.
3. The user can re-arrange the fields they have added by dragging them by a handle on the right side of the field (this has proven to reduce accidental dragging, and be more mobile-friendly than having the entire field be draggable).
4. Only after the user has selected all of the fields for their form and sorted them to their liking will the form get sent to the server for creation/updating.

In short:
1. The database needs a new field, `reports_fields.index`, and the `ReportField` model needs to become a first-class model with this field on it.
2. We don't need (and really shouldn't try to create) new routes just for this. It is part of the creation/editing process, and so it belongs in those existing routes.
3. We don't want a Multi-Step Flow ("second screen"), and will instead modify the `reports/create` and `reports/edit` views to enable this feature.
4. The UIs for reports in general are getting a huge overhaul (this included), and will be a much smoother, JavaScript-based experience.

**Comment by gorel** [comment]
Fair chance this will be moved to the backlog to see if they actually want it. Also, with it getting so close to presentation time, it's more important to get other functionality working than implementing this (rather large) feature.
